spark3/src/main/java/org/apache/iceberg/spark/source/SparkBatchScan.java

```diff
@@ -208,6 +213,33 @@ public Statistics estimateStatistics() {
     return new Stats(sizeInBytes, numRows);
   }
 
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    SparkBatchScan that = (SparkBatchScan) o;
+    return table.toString().equals(that.table.toString()) &&
+        readSchema().equals(that.readSchema()) && // compare Spark schemas to ignore field ids
+        filterExpressions.toString().equals(that.filterExpressions.toString()) &&
```

Comment:
would it make sense to use deepEquals here instead of constructing a string?

Reply:
This is definitely something I think we should improve.

Right now, we don't implement `equals` because we don't want people to use it to test semantic equivalence. Here, we actually want semantic equivalence, but with no utility to evaluate it the simplest way to check for identical structure is the string representation. A better first solution would be to build a comparison utility to test whether two expressions are structurally identical, and after that to introduce one for semantic equivalence.